File: sol/1026/1026.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct Solution {
    void ancestors(TreeNode *node, int ancestorMin, int ancestorMax, int &res){
        if (node == nullptr){
            return;
        }
        res = max(res, max(abs(node->val - ancestorMin), abs(node->val - ancestorMax)));
        ancestorMin = min(ancestorMin, node->val);
        ancestorMax = max(ancestorMax, node->val);
        this->ancestors(node->left, ancestorMin, ancestorMax, res);
        this->ancestors(node->right, ancestorMin, ancestorMax, res);
    }
    int maxAncestorDiff(TreeNode* root) {
        int res = 0;
        this->ancestors(root, root->val, root->val, res);
        return res;
    }
};

File: sol/1026/1026.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        res = 0

        def ancestorMinMax(node, ancestors):
            nonlocal res
            res = max(res, abs(node.val - ancestors[0]), abs(node.val - ancestors[1]))
            prev = ancestors.copy()
            ancestors[0] = min(ancestors[0], node.val)
            ancestors[1] = max(ancestors[1], node.val)
            if node.left is not None:
                ancestorMinMax(node.left, ancestors)
            if node.right is not None:
                ancestorMinMax(node.right, ancestors)
            ancestors[:] = prev[:]

        ancestorMinMax(root, [root.val, root.val])
        return res
